save-orchestrator/src/test/kotlin/org/cqfn/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt

```diff
@@ -216,6 +218,71 @@ class HeartbeatControllerTest {
+        testHeartbeat(
+            agentStatusDtos = listOf(
+                AgentStatusDto(LocalDateTime.now(), AgentState.STARTING, "test-1"),
+                AgentStatusDto(LocalDateTime.now(), AgentState.BUSY, "test-2"),
+            ),
+            heartbeats = listOf(
+                Heartbeat("test-1", AgentState.STARTING, ExecutionProgress(0)),
+                Heartbeat("test-2", AgentState.BUSY, ExecutionProgress(0)),
+                Heartbeat("test-1", AgentState.BUSY, ExecutionProgress(0)),
+                Heartbeat("test-1", AgentState.BUSY, ExecutionProgress(0)),
+                Heartbeat("test-1", AgentState.BUSY, ExecutionProgress(0)),
+                Heartbeat("test-1", AgentState.BUSY, ExecutionProgress(0)),
+            ),
+            heartBeatInterval = 1_000,
+            testBatch = TestBatch(
+                listOf(
+                    TestDto("/path/to/test-1", "WarnPlugin", 1, "hash1", listOf("tag")),
+                    TestDto("/path/to/test-2", "WarnPlugin", 1, "hash2", listOf("tag")),
+                    TestDto("/path/to/test-3", "WarnPlugin", 1, "hash3", listOf("tag")),
+                ),
+                mapOf(1L to "")
+            ),
+            testSuite = TestSuite(TestSuiteType.PROJECT, "", null, null, LocalDateTime.now(), ".", ".").apply {
+                id = 0
+            },
+            mockAgentStatuses = false,
+        ) {
+            // FixMe: we actually need to check the size of crashed agents list somehow
+            verify(dockerService, atLeast(1)).stopAgents(any())
+        }
+    }
+
+    @Test
+    fun `should shutdown all agents, since all of them don't sent heartbeats for some time`() {
+        val agentStatusDtos = listOf(
+            AgentStatusDto(LocalDateTime.now(), AgentState.STARTING, "test-1"),
+            AgentStatusDto(LocalDateTime.now(), AgentState.BUSY, "test-2"),
+        )
+        testHeartbeat(
+            agentStatusDtos = agentStatusDtos,
+            heartbeats = listOf(
+                Heartbeat("test-1", AgentState.STARTING, ExecutionProgress(0)),
+                Heartbeat("test-2", AgentState.BUSY, ExecutionProgress(0)),
+            ),
+            heartBeatInterval = 3000,
+            testBatch = TestBatch(
+                listOf(
+                    TestDto("/path/to/test-1", "WarnPlugin", 1, "hash1", listOf("tag")),
+                    TestDto("/path/to/test-2", "WarnPlugin", 1, "hash2", listOf("tag")),
+                    TestDto("/path/to/test-3", "WarnPlugin", 1, "hash3", listOf("tag")),
+                ),
+                mapOf(1L to "")
+            ),
+            testSuite = TestSuite(TestSuiteType.PROJECT, "", null, null, LocalDateTime.now(), ".", ".").apply {
+                id = 0
+            },
+            mockAgentStatuses = false,
+        ) {
+            // FixMe: we actually need to check the size of crashed agents list somehow
+            verify(dockerService, atLeast(2)).stopAgents(any())
```

Comment:
There should be a matcher for exactly 2 in mockito instead of `atLeast`

Reply:
Actually, the situation is quite more complicated
Since we are not actually start containers, they couldn't be stopped by `stopAgent`, and thus the `crashedAgentsList` won't be cleared, so `stopAgent` method will try to stop them time after time, so it's not really good check at all

It would be good to have access to the `crashedAgentsList` from tests and just check the size of it, but for now I don't know, that it could be possible

I will think more about more appropriate assertsions in these tests

Reply:
Changed to assertion, which check `crashedAgentsList` size, previously, probably because of cache, it didn't worked for me